### PR TITLE
r.i.p old settings

### PR DESCRIPTION
**Pull request by beraldoleal**
@clebergnu We still have some references to this settings.py on avocado-vt, but besides that, this is ready.

### CODE REVIEW

File: avocado/core/parser.py [hunk 1]
```diff
@@ -24,7 +24,6 @@
 
 from . import exit_codes
 from . import varianter
-from . import settings
 from .future.settings import settings as future_settings
 from .future.settings import ConfigFileNotFound, SettingsError
```
Comment (pevogam) [inline avocado/core/parser.py:28]: I guess with this final step you also mean to rename `future_settings` back into `settings`?

Comment (beraldoleal) [reply avocado/core/parser.py:28]: That is right, the PR is almost ready for pushing :)

**Comment by beraldoleal** [comment]
> Hi @beraldoleal, do you mind pointing me to a file documenting the new settings config file? I guess the previous blueprint should now be removed or reused as documentation about the new state of art in the settings? (i.e. what the new settings are as compared to what they should be which was the phrasing in the blueprint)

Hi @pevogam , actually the new settings config file will be the same. We just don't ship the template for now, since that is not complete and the options don't need to be there anymore (once that default values are not got from config file anymore). For now, we can have an empty config file and Avocado will still be functional.

The blueprint (BP001) has a few more stuff to finish, because of that, it is still a working in progress. I will change-it to "implemented" when we close all issues related to the blueprint.

I tried to review our current settings documentation and improve/adapt for the new model, and it was already merged. You can visit here:

https://avocado-framework.readthedocs.io/en/latest/guides/user/chapters/configuring.html

For sure, there are room for improvement.

Please let me know if I'm missing something and where we could improve.

**Comment by beraldoleal** [comment]
@pevogam Ohhh and besides that, you can see **all** available options with `avocado config reference` command (that we will include the output of this command in our "Configuration Reference Guide").

**Comment by clebergnu** [comment]
@beraldoleal could you send a v2 of this PR including this commit: https://github.com/clebergnu/avocado/commit/aa0c8e62c4dddf629ff7f1dfa122e5f1f8fcc6e3

Which I worked on during the Avocado-VT compatibility changes... there's already a lot to cope with there, so I need a stable location for the settings module.

**Comment by willianrampazzo** [comment]
> @beraldoleal could you send a v2 of this PR including this commit: [clebergnu@aa0c8e6](https://github.com/clebergnu/avocado/commit/aa0c8e62c4dddf629ff7f1dfa122e5f1f8fcc6e3)
> 
> Which I worked on during the Avocado-VT compatibility changes... there's already a lot to cope with there, so I need a stable location for the settings module.

I took a quick look at the commit and noticed that `__init__.py` is the only place we are calling the settings object `stgs`. Is it possible to move that to `settings` just for standardization?

**Comment by clebergnu** [comment]
> > @beraldoleal could you send a v2 of this PR including this commit: [clebergnu@aa0c8e6](https://github.com/clebergnu/avocado/commit/aa0c8e62c4dddf629ff7f1dfa122e5f1f8fcc6e3)
> > Which I worked on during the Avocado-VT compatibility changes... there's already a lot to cope with there, so I need a stable location for the settings module.
> 
> I took a quick look at the commit and noticed that `__init__.py` is the only place we are calling the settings object `stgs`. Is it possible to move that to `settings` just for standardization?

I wish we could, but I learned it the hard way.  What happens is, because we're importing "settings" into "avocado.core", it causes a conflict with "avocado.core.settings", which is the module and not the `Settings` instance.

I think this deserves, but I don't it's urgent, moving stuff out of `__init__` or some other change.

**Comment by willianrampazzo** [comment]
> > > @beraldoleal could you send a v2 of this PR including this commit: [clebergnu@aa0c8e6](https://github.com/clebergnu/avocado/commit/aa0c8e62c4dddf629ff7f1dfa122e5f1f8fcc6e3)
> > > Which I worked on during the Avocado-VT compatibility changes... there's already a lot to cope with there, so I need a stable location for the settings module.
> > 
> > 
> > I took a quick look at the commit and noticed that `__init__.py` is the only place we are calling the settings object `stgs`. Is it possible to move that to `settings` just for standardization?
> 
> I wish we could, but I learned it the hard way. What happens is, because we're importing "settings" into "avocado.core", it causes a conflict with "avocado.core.settings", which is the module and not the `Settings` instance.

Ouch!

> 
> I think this deserves, but I don't it's urgent, moving stuff out of `__init__` or some other change.

Okay, we can think about it later, it is not urgent.

**Comment by beraldoleal** [comment]
Thanks @clebergnu , just force-pushed here with your commit.

File: avocado/core/parser.py (hunk 1 above)
Comment (beraldoleal) [reply avocado/core/parser.py]: @pevogam just FYI, I force pushed here with the @clebergnu's commit that is moving the future.settings to settings.

Comment (pevogam) [reply avocado/core/parser.py]: noted, thanks!

**Comment by pevogam** [comment]
Hi @beraldoleal, do we really have to delete the `avocado/etc/avocado/conf.d/README` file? We can definitely get rid of the extra hardcoded defaults but the README file and `conf.d` folder remain a valid location to place configs and it doesn't harm to keep an empty folder with the README file explaining this.

**Comment by beraldoleal** [comment]
> Hi @beraldoleal, do we really have to delete the `avocado/etc/avocado/conf.d/README` file? We can definitely get rid of the extra hardcoded defaults but the README file and `conf.d` folder remain a valid location to place configs and it doesn't harm to keep an empty folder with the README file explaining this.

Yes, I can't see any problem with keeping a README, right now. I will bring the README back if there are no objections.

File: avocado/core/settings.py [hunk 2]
```diff
@@ -58,223 +68,551 @@ def __str__(self):
...
+
+    @property
+    def key(self):
+        return self.namespace.split('.')[-1]
+
+    @property
+    def value(self):
+        if self._value is not None:
+            return self._value
+        return self.default
```
[the remark this answers is not in the document]
Comment (beraldoleal) [reply avocado/core/settings.py:127]: Thanks for this, but this was already there. This PR it is just moving things from one place to another. I will consider this refactoring in a future pull request.

File: avocado/core/settings.py [hunk 3]
```diff
@@ -58,223 +68,551 @@ def __str__(self):
...
+
+        self._update_argparser()
+
+    @property
+    def action(self):
+        if self.key_type is bool:
+            # action is automatic when using bool types
+            if self.default is False:
+                return 'store_true'
+            else:
```
[the remark this answers is not in the document]
Comment (beraldoleal) [reply avocado/core/settings.py:111]: Thanks for this, but this was already there. This PR it is just moving things from one place to another. I will consider this refactoring in a future pull request.

File: avocado/core/settings.py [hunk 4]
```diff
@@ -58,223 +68,551 @@ def __str__(self):
...
+                'default': None}
+
+        if self.nargs:
+            args['nargs'] = self.nargs
+        if self.metavar:
+            args['metavar'] = self.metavar
+        if self.choices:
+            args['choices'] = self.choices
+        if self.action:
+            args['action'] = self.action
```
[the remark this answers is not in the document]
Comment (beraldoleal) [reply avocado/core/settings.py:169]: Thanks for this, but this was already there. This PR it is just moving things from one place to another. I will consider this refactoring in a future pull request.

File: avocado/core/settings.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (YongxueHong) [reply avocado/core/settings.py:127]: Oh~~got it, I misunderstand this goal of PR, sorry for the inconvenience.

**Comment by pevogam** [comment]
Hi @beraldoleal I am currently running from this branch and after adapting the configuration of the plugin on our side don't seem to get colors in the output log file despite having
```
    runner.output.colored                    True
    runner.output.color                      always
```
in the final config (also confirmed in the `job.log` from the test run). Looking at the `output.py` module I suspect the reason could be the time at which the setting is registered. For instance, if I do a simple

```diff
--- a/avocado/core/output.py
+++ b/avocado/core/output.py
@@ -79,6 +79,7 @@ class TermSupport:
         config = settings.as_dict()
         colored = config.get('runner.output.colored')
         force_color = config.get('runner.output.color')
+        LOG_UI.critical(f"{colored} and {force_color}")
         if force_color == "never":
             self.disable()
         elif force_color == "auto":
```

I get three different stages where the class is initiated

```
True and auto
/usr/lib/python3.8/site-packages/avocado/plugins/run.py:283: FutureWarning: The following arguments will be changed to boolean soon: sysinfo, output-check, failfast and keep-tmp. 
  warnings.warn("The following arguments will be changed to boolean soon: "
True and always
True and always
True and always
```

and in the module it seems that there is a central TERM_SUPPORT attribute:

```python
TERM_SUPPORT = TermSupport()
```

So in short, have you managed to obtained a colored log file?

Otherwise I haven't obtained any other errors so far. Good job!

**Comment by beraldoleal** [comment]
> So in short, have you managed to obtained a colored log file?

Hi @pevogam, when you say "log file" are you talking about job.log? Or the std output? My std output is printing colored results.

**Comment by beraldoleal** [comment]
@clebergnu FYI: this was rebased today (a huge one).

**Comment by pevogam** [comment]
> > So in short, have you managed to obtained a colored log file?
> 
> Hi @pevogam, when you say "log file" are you talking about job.log? Or the std output? My std output is printing colored results.

No, for instance doing "avocado run > some.log" to store the current statuses in the long term. This had colors working before but no longer now.

> @clebergnu FYI: this was rebased today (a huge one).

I could also rerun with the rebased code although I don't think it will make any difference.

**Comment by beraldoleal** [comment]
> > > So in short, have you managed to obtained a colored log file?
> > 
> > 
> > Hi @pevogam, when you say "log file" are you talking about job.log? Or the std output? My std output is printing colored results.
> 
> No, for instance doing "avocado run > some.log" to store the current statuses in the long term. This had colors working before but no longer now.

Hum, so yes, from the Avocado POV this is the output. And yes, it is working for me, strange. 

> > @clebergnu FYI: this was rebased today (a huge one).
> 
> I could also rerun with the rebased code although I don't think it will make any difference.

Yeah, I don't think this will fix I just changed the imports on multiple files.

**Comment by pevogam** [comment]
> Hum, so yes, from the Avocado POV this is the output. And yes, it is working for me, strange.

So you can confirm that following the minimal step `avocado config > some.log`, there will be a blue color in the "Config files read..." line in the output of `head -n 5 some.log`?

You might also need the settings

```
$ avocado config | grep color
    runner.output.colored                    True
    runner.output.color                      always
```

**Comment by beraldoleal** [comment]
@pevogam I think that I got now. I will open an issue for that since that is related to the order as you noticed.

**Comment by clebergnu** [comment]
@pevogam found an issue that may require an reorder of the plugin initialization and the merging of the settings with the configs.

Given that this has been thoroughly tested by, among other things, the Avocado-VT changes, I'm going to rebase and merge this -- it'll be easier to address the smaller issues that way.

**Comment by clebergnu** [comment]
Merged a rebased version!
